extensions/amp-story/1.0/amp-story-affiliate-link.js

```diff
@@ -0,0 +1,140 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Affiliate link component that expands when clicked.
+ */
+
+import {StateProperty, getStoreService} from './amp-story-store-service';
+import {htmlFor} from '../../../src/static-template';
+
+/**
+ * Links that are affiliate links.
+ * @const {string}
+ */
+export const AFFILIATE_LINK_SELECTOR = 'a[data-affiliate-link-icon]';
+
+/**
+ * Custom property signifying a built link.
+ * @const {string}
+ */
+export const AFFILIATE_LINK_BUILT = '__AMP_AFFILIATE_LINK_BUILT';
+
+export class AmpStoryAffiliateLink {
+  /**
+   * @param {!Window} win
+   * @param {!Element} element
+   */
+  constructor(win, element) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Element} */
+    this.element_ = element;
+
+    /** @private {?Element} */
+    this.textEl_ = null;
+
+    /** @private {?Element} */
+    this.launchEl_ = null;
+
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = getStoreService(this.win_);
+  }
+
+  /**
+   * Builds affiliate link.
+   */
+  build() {
+    if (this.element_[AFFILIATE_LINK_BUILT]) {
+      return;
+    }
+
+    this.addPulseElement_();
+    this.addIconElement_();
+    this.addText_();
+    this.addLaunchElement_();
```

Comment:
Since we're writing in the DOM, can we wrap all the previous calls into a `mutate` task to avoid layout thrashing?